x/emissions/module/abci.go

```diff
@@ -90,5 +91,32 @@ func EndBlocker(ctx context.Context, am AppModule) error {
 	}
 	wg.Wait()
 
+	var wgAddingNonces sync.WaitGroup
+	// Add unfulfilled worker nonces
+	churnReadyTopics, err := am.keeper.GetChurnReadyTopics(sdkCtx)
+	if err != nil {
+		fmt.Println("Error getting active topics and met demand: ", err)
+	} else {
+		if len(churnReadyTopics.Topics) == 0 {
+			fmt.Println("No churn ready topics.")
+		} else {
+			for _, topic := range churnReadyTopics.Topics {
+				wgAddingNonces.Add(1)
+				go func(topic *emissionstypes.Topic) {
+					defer wgAddingNonces.Done()
+					if blockNumber == topic.EpochLastEnded+topic.EpochLength {
+						nextNonce := emissionstypes.Nonce{BlockHeight: blockNumber + topic.EpochLength}
+						err = am.keeper.AddWorkerNonce(sdkCtx, topic.Id, &nextNonce)
+						if err != nil {
+							fmt.Println("Error adding worker nonce: ", err)
+							return
+						}
+					}
+				}(topic)
+			}
+		}
+	}
+	wgAddingNonces.Wait()
```

Comment:
there's no need to query churnReadyTopics again + iterate through then. You already have everything available above in `topTopicsActiveWithDemand`. I guess you just need add the `AddWorkerNonce` right below `UpdateTopicEpochLastEnded`.

Reply:
Good catch, changing it now.